drivers/base/auxiliary.c

```diff
@@ -195,6 +195,52 @@ int __auxiliary_device_add(struct auxiliary_device *auxdev, const char *modname)
 }
 EXPORT_SYMBOL_GPL(__auxiliary_device_add);
 
+/**
+ * __auxiliary_device_register - register an auxiliary bus device
+ * @auxdev: auxiliary bus device to add to the bus
+ * @auxdev_container: pointer to structure containing @auxdev
+ * @modname: name of the parent device's driver module
+ *
+ * This routine combines the first (init) and second (add) steps together.
+ * The header is different from traditional _register function. The auxiliary
+ * device is typically embedded in a larger structure, and we need to free that
+ * larger structure if @auxiliary_device_init fails.
+ *
+ * All errors are handled internally, i.e. kfree is called after a failure of
+ * @auxiliary_device_init() and auxiliary_device_uninit() is called if
+ * @auxiliary_device_add fails. The container is assumed to be allocated
+ * normally, i.e. not be device-managed.
+ *
+ * The expectation is that users will call the "auxiliary_device_register" macro
+ * so that the caller's KBUILD_MODNAME is automatically inserted for the modname
+ * parameter.  Only if a user requires a custom name would this version be
+ * called directly.
+ */
+int __auxiliary_device_register(struct auxiliary_device *auxdev,
+				void *auxdev_container,
+				const char *modname)
+{
+	int ret;
+
+	if (!auxdev || !auxdev_container)
+		return -EINVAL;
```

Comment:
@plbossart I feel like this is a bit of an overkill. Why would someone pass a NULL pointer for a device they want to register? I think we could skip this step, and call device_initialize(auxdev->dev) as the first call. The code for device_initialize() clearly states:
```
 * All fields in @dev must be initialized by the caller to 0, except
 * for those explicitly set to some other value.  The simplest
 * approach is to use kzalloc() to allocate the structure containing
 * @dev.
```
And once the device is initialized, then every error can simply be handled with the uninit()

Reply:
I think the aux device should be oblivious to how it is allocated and its container. Handling the container in the auxbus code seem wrong to me.